.github/workflows/ibis-backends.yml

```diff
@@ -425,3 +425,90 @@ jobs:
         with:
           name: ${{ matrix.backend }}-${{ matrix.python-version }}
           path: junit.xml
+
+  test_datafusion:
+    name: DataFusion ${{ matrix.datafusion-version }} ubuntu-latest python-${{ matrix.python-version }}
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        datafusion-version:
+          - master
+          - latest
+        python-version:
+          - "3.7"
+          - "3.9"
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+
+      - name: checkout
+        uses: actions/checkout@v2
+        if: ${{ matrix.datafusion-version == 'master' }}
+        with:
+          path: datafusion
+          repository: apache/arrow-datafusion
+
+      - name: install rust
+        uses: actions-rs/toolchain@v1
+        if: ${{ matrix.datafusion-version == 'master' }}
+        with:
+          profile: minimal
+          toolchain: stable
+
+      - name: install python
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: cache dependencies
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/pypoetry
+          key: datafusion-${{ runner.os }}-${{ matrix.python-version }}-${{ hashFiles('poetry.lock') }}
```

Comment:
You need to cache with the specific dependency coming out of the `install python` step to ensure the cache is invalidated for minor Python version upgrades. We do this in `ibis-main.yml`. Check there for examples.

Reply:
Following up here, you need to assign the `setup-python` step an `id` (like `install_python`) and then use `${{ steps.install_python.outputs.python-version }}` instead of `${{ matrix.python-version }}` here, otherwise the cache won't be invalidated when a new patch version of Python shows up (e.g., 3.9.8 -> 3.9.9) and this will break in a hard-to-debug way.